### PR TITLE
Add a Reactive Messaging HTTP extension

### PR DESCRIPTION
**Pull request by michalszynkiewicz**

### CODE REVIEW

File: extensions/reactive-messaging-http/deployment/src/main/java/io/quarkus/reactivemessaging/http/ReactiveHttpProcessor.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import io.quarkus.reactivemessaging.http.config.WebsocketStreamConfig;
+import io.quarkus.vertx.http.deployment.RouteBuildItem;
+import io.vertx.core.Handler;
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * @author Michal Szynkiewicz, michal.l.szynkiewicz@gmail.com
+ *         <br>
+ *         Date: 28/08/2019
+ */
```
Comment (gsmet) [inline extensions/reactive-messaging-http/deployment/src/main/java/io/quarkus/reactivemessaging/http/ReactiveHttpProcessor.java]: Same as usual, could you get rid of all the author tags? Better use git to track that. 

Thanks.

**Review by gsmet — CHANGES_REQUESTED** [review]
Just a minor comment, I’ll let the reactive people check the code.

File: extensions/reactive-messaging-http/runtime/src/main/java/io/quarkus/reactivemessaging/http/config/ReactiveHttpConfig.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import io.quarkus.reactivemessaging.http.QuarkusWebsocketConnector;
+
+/**
+ * ms@author Michal Szynkiewicz, michal.l.szynkiewicz@gmail.com
+ * <br>
+ * Date: 27/09/2019
+ */
+@Singleton
+public class ReactiveHttpConfig {
+    private static final String CONNECTOR = ".connector";
```
Comment (gsmet) [inline extensions/reactive-messaging-http/runtime/src/main/java/io/quarkus/reactivemessaging/http/config/ReactiveHttpConfig.java]: All the non API classes should be in a runtime subpackage per convention.

Comment (michalszynkiewicz) [reply extensions/reactive-messaging-http/runtime/src/main/java/io/quarkus/reactivemessaging/http/config/ReactiveHttpConfig.java]: should be okay now

File: extensions/reactive-messaging-http/deployment/src/main/java/io/quarkus/reactivemessaging/http/ReactiveHttpProcessor.java (hunk 1 above)
Comment (michalszynkiewicz) [reply extensions/reactive-messaging-http/deployment/src/main/java/io/quarkus/reactivemessaging/http/ReactiveHttpProcessor.java]: I really, really, really like my author tags :(

**Comment by michalszynkiewicz** [comment]
@cescoffier if you have a while, could you take a look and add some comment what you think and if it's worth continuing with it now.

File: extensions/reactive-messaging-http/deployment/src/main/java/io/quarkus/reactivemessaging/http/ReactiveHttpProcessor.java (hunk 1 above)
Comment (gsmet) [reply extensions/reactive-messaging-http/deployment/src/main/java/io/quarkus/reactivemessaging/http/ReactiveHttpProcessor.java]: Again, this is project policy. We decided to not include them in our source code (at least, not more of them). AFAICS, there are still some in this patch.

**Comment by gsmet** [comment]
This apparently needs a rebase.

Thanks!

**Comment by michalszynkiewicz** [comment]
ading [stale] label back :)

**Comment by gsmet** [comment]
This needs a rebase and a `@author` tag cleanup. Then @cescoffier should decide if we want it and review the code.

File: extensions/reactive-messaging-http/runtime/src/main/java/io/quarkus/reactivemessaging/http/runtime/HttpMessage.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    private final Buffer buffer;
+
+    public HttpMessage(Buffer buffer) {
+        this.buffer = buffer;
+    }
+
+    @Override
+    public Buffer getPayload() {
+        return buffer;
+    }
```
Comment (cescoffier) [inline extensions/reactive-messaging-http/runtime/src/main/java/io/quarkus/reactivemessaging/http/runtime/HttpMessage.java]: What about headers, url...?

File: extensions/reactive-messaging-http/runtime/src/main/java/io/quarkus/reactivemessaging/http/runtime/QuarkusWebsocketConnector.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    @Inject
+    ReactiveHttpHandlerBean handlerBean;
+
+    @Override
+    public PublisherBuilder<HttpMessage> getPublisherBuilder(Config config) {
+        String path = config.getOptionalValue("path", String.class)
+                .orElseThrow(() -> new IllegalArgumentException("The `path` must be set"));
+
+        BehaviorProcessor<HttpMessage> processor = handlerBean.getWebsocketProcessor(path);
+        return ReactiveStreams.fromPublisher(processor);
```
Comment (cescoffier) [inline extensions/reactive-messaging-http/runtime/src/main/java/io/quarkus/reactivemessaging/http/runtime/QuarkusWebsocketConnector.java]: Any reason (except time) to not do the other direction?

File: extensions/reactive-messaging-http/runtime/src/main/java/io/quarkus/reactivemessaging/http/runtime/ReactiveHttpHandlerBean.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    private final Map<String, BehaviorProcessor> processors = new HashMap<>();
+    private final Map<String, BehaviorProcessor> websocketProcessors = new HashMap<>();
+
+    public void handleHttp(RoutingContext event) {
+        String path = event.normalisedPath();
+        HttpMethod method = event.request().method();
+        BehaviorProcessor processor = processors.get(key(path, method));
+        if (processor != null) {
+            processor.onNext(new HttpMessage(event.getBody()));
```
Comment (cescoffier) [inline extensions/reactive-messaging-http/runtime/src/main/java/io/quarkus/reactivemessaging/http/runtime/ReactiveHttpHandlerBean.java]: Here I would create a Message with more metadata, maybe even passing the RoutingContext (that could be retrieved using an `unwrap` method)

File: extensions/reactive-messaging-http/runtime/src/main/java/io/quarkus/reactivemessaging/http/runtime/ReactiveHttpHandlerBean.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        }
+    }
+
+    public void handleWebsocket(RoutingContext event) {
+        String path = event.normalisedPath();
+        BehaviorProcessor processor = websocketProcessors.get(path);
+        HttpMethod method = event.request().method();
+        if (processor != null) {
+            event.request().upgrade().handler(b -> {
+                processor.onNext(new HttpMessage(b));
```
Comment (cescoffier) [inline extensions/reactive-messaging-http/runtime/src/main/java/io/quarkus/reactivemessaging/http/runtime/ReactiveHttpHandlerBean.java]: WebSocket is not really HTTP, I would create a specific `WebSocketMessage`

File: extensions/reactive-messaging-http/runtime/src/main/java/io/quarkus/reactivemessaging/http/runtime/config/ReactiveHttpConfig.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+import io.quarkus.reactivemessaging.http.runtime.QuarkusHttpConnector;
+import io.quarkus.reactivemessaging.http.runtime.QuarkusWebsocketConnector;
+
+/**
+ * ms@author Michal Szynkiewicz, michal.l.szynkiewicz@gmail.com
+ * <br>
+ * Date: 27/09/2019
+ */
+@Singleton
```
Comment (cescoffier) [inline extensions/reactive-messaging-http/runtime/src/main/java/io/quarkus/reactivemessaging/http/runtime/config/ReactiveHttpConfig.java]: Any reason for this singleton?

**Review by cescoffier — CHANGES_REQUESTED** [review]

**Comment by kunal-kushwaha** [comment]
@michalszynkiewicz : Can I rebase it and bring it up to date and submit a new PR if it's not a WIP? Thanks

**Comment by michalszynkiewicz** [comment]
@kunal-kushwaha I'm on some more important tasks for a long time, planned to get back to it when I'm finished with them. But if you'd like to pick it up and continue, that would be great :)

There is a lot of things to be done in this PR before it's ready.
You can take a look at https://github.com/quarkusio/quarkus/issues/3044 for what is expected in it.

If you'd like to help with it, please ping me on Zulip.

**Comment by gsmet** [comment]
I'm closing this one as out of date.

**Comment by michalszynkiewicz** [comment]
:+1:  I just came back to this work but will create a new PR

**Pull request closed without merging**